streaming/src/main/scala/org/apache/spark/streaming/util/BatchedWriteAheadLog.scala

```
@@ -0,0 +1,206 @@
+ * de-aggregation after the `read` method. In addition, the `WriteAheadLogRecordHandle` returned
+ * after the write will contain the batch of records rather than individual records.
+ *
+ * When writing a batch of records, the `time` passed to the `wrappedLog` will be the timestamp
+ * of the latest record in the batch. This is very important in achieving correctness. Consider the
+ * following example:
+ * We receive records with timestamps 1, 3, 5, 7. We use "log-1" as the filename. Once we receive
+ * a clean up request for timestamp 3, we would clean up the file "log-1", and lose data regarding
+ * 5 and 7.
+ *
+ * All other methods of the WriteAheadLog interface will be passed on to the wrapped WriteAheadLog.
+ */
+private[util] class BatchedWriteAheadLog(val wrappedLog: WriteAheadLog)
+  extends WriteAheadLog with Logging {
+
+  import BatchedWriteAheadLog._
+
+  // exposed for tests
+  private val walWriteQueue = new LinkedBlockingQueue[Record]()
+
+  // Whether the writer thread is active
+  @volatile private var active: Boolean = true
+  private val buffer = new ArrayBuffer[Record]()
+
+  private val batchedWriterThread = startBatchedWriterThread()
+
+  /**
+   * Write a byte buffer to the log file. This method adds the byteBuffer to a queue and blocks
+   * until the record is properly written by the parent.
+   */
+  override def write(byteBuffer: ByteBuffer, time: Long): WriteAheadLogRecordHandle = {
+    val promise = Promise[WriteAheadLogRecordHandle]()
+    walWriteQueue.offer(Record(byteBuffer, time, promise))
+    Await.result(promise.future, WAL_WRITE_STATUS_TIMEOUT.milliseconds)
+  }
+
+  /**
+   * Read a segment from an existing Write Ahead Log. The data may be aggregated, and the user
+   * should de-aggregate using [[BatchedWriteAheadLog.deaggregate]]
+   *
+   * This method is handled by the parent WriteAheadLog.
+   */
+  override def read(segment: WriteAheadLogRecordHandle): ByteBuffer = {
+    wrappedLog.read(segment)
```

Maybe throw an exception here? E.g., `throw new UnsupportedOperationException("BatchedWriteAheadLog.read")`. To make the error more explicit if BatchedWriteAheadLog is used in Executor by mistake.